### PR TITLE
74 acm ssl integration

### PR DESCRIPTION
**Pull request by Biplob14**
- [x] Does not depend on other PRs
- [ ] Depends on other PR being merged first: #

## What kind of change does this PR introduce (check at least one)

- [ ] Bugfix
- [ ] Improvement
- [x] New feature
- [ ] Refactor
- [ ] Infrastructure
- [ ] Other (please specify)

## How would you describe the changes in a changelog (1-2 lines)?
- SSL certificate implementation on ACM to route53 domain

## How to test

- [ ] Needs regression testing (add QA needed label)

## Checklist

- [x] I have tested these changes locally and they function as expected.
- [ ] I have added or updated necessary documentation to reflect the changes.
- [x] I have followed the project's coding style and guidelines.
- [ ] I have added appropriate test coverage to ensure the changes are functioning correctly.
- [x] I have checked for any potential conflicts with other branches or pull requests.
- [ ] I have updated necessary OpenAPI specs

## Deploy risk & follow up

- [x] No risk
- [ ] Low
- [ ] High

### CODE REVIEW

File: deployment/aws/terraform/main.tf [hunk 1]
```diff
@@ -1,4 +1,9 @@
 module "aws" {
-  source              = "../../../iac/terraform/aws"
-  environment         = var.aws_environment
+  source         = "../../../iac/terraform/aws"
+  environment    = var.aws_environment
+  hosted_zone_id = data.aws_route53_zone.route53_zone.zone_id
+}
+
+data "aws_route53_zone" "route53_zone" {
+  name = "bloodconnect.net"
```
Comment (melon-ruet) [inline deployment/aws/terraform/main.tf:8]: Pass it from terraform var

File: deployment/localstack/terraform/main.tf [hunk 2]
```diff
@@ -1,4 +1,11 @@
...
-  environment         = "localstack"
+  source         = "../../../iac/terraform/aws"
+  environment    = "localstack"
+  hosted_zone_id = aws_route53_zone.my_zone.zone_id
+}
+
+resource "aws_route53_zone" "my_zone" {
+  #checkov:skip=CKV2_AWS_38: "Ensure Domain Name System (DNS) query logging is enabled for Amazon Route 53 hosted zones"
+  #checkov:skip=CKV2_AWS_39: "Ensure Domain Name System Security Extensions (DNSSEC) signing is enabled for Amazon Route 53 public hosted zones"
+  name = "bloodconnect.net"
```
Comment (melon-ruet) [inline deployment/localstack/terraform/main.tf:10]: Is it working for localstack?

File: deployment/localstack/terraform/main.tf [hunk 3]
```diff
@@ -1,4 +1,11 @@
...
-  source              = "../../../iac/terraform/aws"
-  environment         = "localstack"
+  source         = "../../../iac/terraform/aws"
+  environment    = "localstack"
+  hosted_zone_id = aws_route53_zone.my_zone.zone_id
+}
+
+resource "aws_route53_zone" "my_zone" {
+  #checkov:skip=CKV2_AWS_38: "Ensure Domain Name System (DNS) query logging is enabled for Amazon Route 53 hosted zones"
+  #checkov:skip=CKV2_AWS_39: "Ensure Domain Name System Security Extensions (DNSSEC) signing is enabled for Amazon Route 53 public hosted zones"
```
Comment (melon-ruet) [inline deployment/localstack/terraform/main.tf:9]: Why DNSSEC signing is skipped?

File: iac/terraform/aws/route53/variables.tf [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+}
+
+variable "domain_validation_method" {
+  default     = "DNS"
+  description = "domain validation method"
+  type        = string
+}
+
+variable "hosted_zone_id" {
+  default     = "Z0872264215II0BIDZ5EC"
```
Comment (melon-ruet) [inline iac/terraform/aws/route53/variables.tf:14]: This is sensitive data. Do not set default. Please clean your commit.

File: iac/terraform/aws/route53/variables.tf [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  type        = string
+}
+
+variable "domain_validation_method" {
+  default     = "DNS"
+  description = "domain validation method"
+  type        = string
+}
+
+variable "hosted_zone_id" {
```
Comment (melon-ruet) [inline iac/terraform/aws/route53/variables.tf:13]: Can you dynamically get it?
https://registry.terraform.io/providers/hashicorp/aws/latest/docs/data-sources/route53_zone

**Review by melon-ruet — CHANGES_REQUESTED** [review]

File: deployment/aws/terraform/main.tf (hunk 1 above)
Comment (Biplob14) [reply deployment/aws/terraform/main.tf:8]: ok

File: deployment/localstack/terraform/main.tf (hunk 2 above)
Comment (Biplob14) [reply deployment/localstack/terraform/main.tf:10]: yes, it was creating a new hosted zone on localstack.

File: iac/terraform/aws/route53/variables.tf (hunk 5 above)
Comment (Biplob14) [reply iac/terraform/aws/route53/variables.tf:13]: yes, It's doable.

File: deployment/localstack/terraform/main.tf (hunk 3 above)
Comment (Biplob14) [reply deployment/localstack/terraform/main.tf:9]: Created another new ticket https://github.com/CraftsmenLtd/BloodConnect/issues/108 related to this problem.

File: iac/terraform/aws/route53/variables.tf (hunk 4 above)
Comment (melon-ruet) [reply iac/terraform/aws/route53/variables.tf]: You need to squash commits

**Comment by melon-ruet** [comment]
Terraform plan is not working

File: iac/terraform/aws/route53/variables.tf (hunk 5 above)
Comment (melon-ruet) [reply iac/terraform/aws/route53/variables.tf:13]: If you get this dynamically, why this variable is necessary?

File: iac/terraform/aws/variables.tf [hunk 6]
```diff
@@ -42,3 +42,8 @@ variable "combined_openapi_file" {
   description = "Combined OpenAPI file"
   default     = "combined-openapi.json"
 }
+
+variable "hosted_zone_id" {
```
Comment (melon-ruet) [inline iac/terraform/aws/variables.tf:46]: Is it necessary?

File: iac/terraform/aws/route53/ssl.tf [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+resource "aws_route53_record" "validate_record" {
+  for_each = {
+    for dvo in aws_acm_certificate.ssl_certificate.domain_validation_options : dvo.domain_name => {
+      name   = dvo.resource_record_name
+      record = dvo.resource_record_value
+      type   = dvo.resource_record_type
+    }
+  }
+
+  zone_id         = var.hosted_zone_id
```
Comment (melon-ruet) [inline iac/terraform/aws/route53/ssl.tf:30]: Can you get it dynamically?

File: deployment/aws/terraform/main.tf [hunk 8]
```diff
@@ -1,4 +1,10 @@
 module "aws" {
-  source              = "../../../iac/terraform/aws"
-  environment         = var.aws_environment
+  source         = "../../../iac/terraform/aws"
+  environment    = var.aws_environment
+  hosted_zone_id = data.aws_route53_zone.route53_zone.zone_id
 }
+
+data "aws_route53_zone" "route53_zone" {
```
Comment (melon-ruet) [inline deployment/aws/terraform/main.tf:7]: If roite53 zone is not necessary in deployment directory, pass domain name to iac directory and move this code there

**Review by melon-ruet — CHANGES_REQUESTED** [review]

**Comment by Biplob14** [comment]
SSL certificate generation is done manually -
- to stop SSL request on `dev` deployment
- multi region development issue

**Pull request closed without merging**